### ISSUE TITLE
[rootless] Naive container -> host networking

### ISSUE THREAD

**Issue by maybe-sybr** [containers/podman#7432]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

I've been trying to do some naive inter-pod networking without having to do the legwork of probing and passing the host IP into my containers. Unfortunately it seems like something is now preventing me from (ab)using slirp gateway/DNS IPs to connect to services listening on the host or in other containers (exposed via published ports in their pod). I think this may have worked in the past prior to podman starting to use the slirp `--disable-host-loopback` option.

My overarching goal is to provide a simple default host to connect to which will make inter-pod networking Just Work (TM) in development environments where it's not practical for me to rely on more robust DNS and load balancing that would happen on something like k8s (maybe noteworthy, I'm using `podman play kube` to bring pods up). This use case sounds like it would be adequately solved by a solution for #6863, but that seems like it'll be a non-trivial change, and for my purposes treating this as a container -> host problem would also be fine and likely simpler to solve/reject.

I'd like to either be able to specify something like the magic slirp gateway IP `10.0.2.2` or some magic DNS which resolves to all non-loopback addresses on the host side so that the container side could attempt to perform client side DNS failover and eventually connect to a service published by the host on a non-loopback address. From a few brief netcat pings between containers, it seems like just DNS answering with all IPs when questioned about a magic hostname would work and not allow any surprising connectivity on my Fedora 32 machine.

**Steps to reproduce the issue:**

1. `(host)$ nc -lvvk 12345

2. `(container)$ nc -vv <10.0.2.2|some_hostname.net.pod> 12345`

**Describe the results you received:**

Connection refused

**Describe the results you expected:**

Connection accepted

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
sh-5.0$ podman version
Version:      2.0.4
API Version:  1
Go Version:   go1.14.6
Built:        Thu Jan  1 10:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
sh-5.0$ podman info --debug
host:
  arch: amd64
  buildahVersion: 1.15.0
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.19-1.fc32.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.19, commit: 5dce9767526ed27f177a8fa3f281889ad509fea7'
  cpus: 8
  distribution:
    distribution: fedora
    version: "32"
  eventLogger: file
  hostname: host
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 300000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 300000
      size: 65536
  kernel: 5.7.16-200.fc32.x86_64
  linkmode: dynamic
  memFree: 3377954816
  memTotal: 8235962368
  ociRuntime:
    name: crun
    package: crun-0.14.1-1.fc32.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.14.1
      commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.4-1.fc32.x86_64
    version: |-
      slirp4netns version 1.1.4
      commit: b66ffa8e262507e37fca689822d23430f3357fe8
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 2
  swapFree: 8392798208
  swapTotal: 8392798208
  uptime: 49m 6.35s
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /home/user/.config/containers/storage.conf
  containerStore:
    number: 3
    paused: 0
    running: 3
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.1.2-1.fc32.x86_64
      Version: |-
        fusermount3 version: 3.9.1
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.9.1
        using FUSE kernel interface version 7.31
  graphRoot: /home/user/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 226
  runRoot: /run/user/1000/containers
  volumePath: /home/user/.local/share/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Thu Jan  1 10:00:00 1970
  GitCommit: ""
  GoVersion: go1.14.6
  OsArch: linux/amd64
  Version: 2.0.4
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.0.4-1.fc32.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical host

**Comment by maybe-sybr** [comment]
Having written that and read #6863 more deeply throughout, I'm fairly convinced that the right thing to be considered here is the introduction of a magic DNS record for the host rather than messing with whatever slirp is doing to make 10.0.2.2 work how I mentioned above. What I'm not familiar enough with is how DNS works in podman and if we already have some system which could do this.

**Comment by mheon** [comment]
DNS could be complicated, but I could do an entry in /etc/hosts trivially

**Comment by maybe-sybr** [comment]
That's a great idea, @mheon. For me, not having it deal with runtime changes to host addresses/etc would be perfectly acceptable. Not sure how to best pick which IP to inject into the container for general cases though but I'm already using `hostname -I | cut -d ' ' -f1` elsewhere in my dev scripts which works fine for `::` and `0.0` listeners and might be good enough.

**Comment by maybe-sybr** [comment]
Another alternative/follow up might be to support hostAliases in podman play, which I could then populate manually using a `sed` prior to `podman play kube` to act as a poor man's helm. That would shift the onus for picking a sensible IP back to the user and presumably the plumbing would be mostly around whatever `--add-host` does for `podman create` and the kube parsing/generation.

**Comment by rhatdan** [comment]
@mheon @maybe-sybr Are you just looking for --add-host?

**Comment by maybe-sybr** [comment]
@rhatdan I think the problem I was having was that things like `--add-host` aren't available via the `podman play kube` CLI. I would have definitely used that and been happy if it jammed the host entry into all containers. This is what happens when I try to do something naive like that:

```
$ podman play kube --add-host some-host.local:1.2.3.4 busybox.yaml
Error: unknown flag: --add-host
$ podman --version
podman version 2.0.6
```

Unless something got added recently which I've missed, I don't think there's a way to do what I was asking for.

*However!* Thinking back on the waffle I've commented above, it seems like a super bad idea to be supporting magic like exposing the host to pods/containers by default. If they aren't given a routable IP for the host, they shouldn't have one (even if it would be possible for them to discover it through evil means).

Having an `--add-host` arg passthrough from `play kube` to the eventual `pod create`s it does seems like a sensible solution.

For the moment, I've been doing some janky stuff with an env-var and deprecating that dev environment with usernetes since being able to use k8s services avoids this issue for me entirely and is more representative of prod. So I'm not blocked or needing this particularly strongly.
